### ISSUE TITLE
'docker ps' and 'docker run' commands never finish, other like 'info' are working

### ISSUE THREAD

**Issue by piontec** [docker/docker#30782]
**Description**

Dockerd became unresponsive at 06:04:21 GMT (see logs). Commands like "docker ps" and "docker run" are never completed within reasonable time (1 min). "docker info" is working. Containers are running and weren't restarted.

**Steps to reproduce the issue:**
Unfortunately, I don't know the way to reproduce the problem.

**Additional information you deem important (e.g. issue happens only occasionally):**
This happened after about 2 weeks of running successfully with similar load.

**Output of `docker version`:**

```
# docker version
Client:
 Version:      1.12.6-cs7
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   681cddc
 Built:        Tue Jan 24 18:01:10 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6-cs7
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   681cddc
 Built:        Tue Jan 24 18:01:10 2017
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
# docker info
Containers: 472
 Running: 355
 Paused: 0
 Stopped: 117
Images: 12929
Server Version: 1.12.6-cs7
Storage Driver: aufs
 Root Dir: /opt/io1/docker/aufs
 Backing Filesystem: extfs
 Dirs: 12143
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host overlay bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-59-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 128
Total Memory: 1.876 TiB
Name: ip-10-69-11-89
ID: UGZS:UFD3:GB4C:W5MX:JU2L:K7PH:6ZWS:4GPM:27Q5:UNNN:X3DC:YDT7
Docker Root Dir: /opt/io1/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 5526
 Goroutines: 4451
 System Time: 2017-02-07T07:56:27.892730562Z
 EventsListeners: 2
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
- running on AWS, x1.32xlarge instance
- restarting the daemon takes very long time
```
# time systemctl restart docker

real    25m39.047s
user    0m0.028s
sys     0m0.012s
```
- after the restart, dockerd was behaving the same
- please see attached syslog from this period (with debug) and thread dump requested when the commands were unresponsive

[2017-02-07-docker.log.txt](https://github.com/docker/docker/files/757208/2017-02-07-docker.log.txt)
[2017-02-07-dockerd_threaddump.log.txt](https://github.com/docker/docker/files/757207/2017-02-07-dockerd_threaddump.log.txt)

**Comment by vdemeester** [comment]
Looking at the info, that's quite a lot of objects (containers and images).

```
Containers: 472
 Running: 355
 Paused: 0
 Stopped: 117
Images: 12929
```

And `Debug Mode (server): true` makes me feel there might a crazy amount of I/O (daemon logs in debug writes a lot !).

There is also some logs that are a little bit worrysome for the filesystem : `Feb  7 06:06:42 ip-10-69-11-89 dockerd[5364]: time="2017-02-07T06:06:42.913974461Z" level=error msg="Handler for POST /v1.24/exec/b1f555823d878e11d4a118aa66a74446935ffb2a5938955b570d76f5d7bf4861/resize returned error: rpc error: code = 2 desc = write /var/run/docker/libcontainerd/containerd/f0e4c98daeba3f7405e31659fd8c06299ee2bf1195979a13487bed7c97d95117/b1f555823d878e11d4a118aa66a74446935ffb2a5938955b570d76f5d7bf4861/control: bad file descriptor"`

/cc @tonistiigi @mlaventure

**Comment by piontec** [comment]
Important update: restarting dockerd helped only for few seconds, then commands like "ps" and "run" became unresponsive again. We restarted dockerd for 2nd time, this time the restart took 39 minutes.
Also, I have some containers, which can be inspected, like
```
root@ip-10-69-11-89:~/tmp# time docker inspect docker-enforcer > /dev/nullreal    0m0.033suser    0m0.012ssys     0m0.020s
```

but for the others, `inspect` never ends:
```
root@ip-10-69-11-89:~/tmp# time docker inspect as > /dev/null                                                                                                                                               
^C

real    0m35.452s
user    0m0.008s
sys     0m0.024s
```

**Comment by vdemeester** [comment]
@piontec what is the load average of your server ? :angel:

**Comment by piontec** [comment]
with 128 CPUs
```
# uptime 
 11:36:35 up 22 days, 21:43, 21 users,  load average: 60.49, 64.13, 63.07
```

**Comment by piontec** [comment]
Since you suggested, that the number of objects we have in system is large, I tried to run some cleanups. Still, I couldn't run any 'docker rm' or 'docker rmi'. I restarted dockerd again and immediately after the restart started to stop 7 heavy logging containers and a lot of images. This time it succeeded and now we got:
```
# docker info
Containers: 407
 Running: 298
 Paused: 0
 Stopped: 109
Images: 744
Server Version: 1.12.6-cs7
Storage Driver: aufs
 Root Dir: /opt/io1/docker/aufs
 Backing Filesystem: extfs
 Dirs: 4391
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-59-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 128
Total Memory: 1.876 TiB
Name: ip-10-69-11-89
ID: UGZS:UFD3:GB4C:W5MX:JU2L:K7PH:6ZWS:4GPM:27Q5:UNNN:X3DC:YDT7
Docker Root Dir: /opt/io1/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 2924
 Goroutines: 1648
 System Time: 2017-02-07T12:54:28.291665011Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Comment by piontec** [comment]
Also, some users tried to create containers when dockerd was restarting. This caused creation of over 100 containers in 'Created' state. When I tried to remove them with `docker rm`, they just changed their status to 'Dead'.
When trying to remove them now, I get the following error for all of them:
```
Error response from daemon: Unable to remove filesystem for 13653abc0560820d2b1ad3ee661ffc4cf54fb73af7f6a46af2fb6ee69778e070: remove /opt/io1/docker/containers/13653abc0560820d2b1ad3ee661ffc4cf54fb73af7f6a46af2fb6ee69778e070/shm: device or resource busy
Error response from daemon: Unable to remove filesystem for f6d822182385260ae037915cd93d2e7087f3e14d0ded6f541bd0319d2adccf54: remove /opt/io1/docker/containers/f6d822182385260ae037915cd93d2e7087f3e14d0ded6f541bd0319d2adccf54/shm: device or resource busy
Error response from daemon: Unable to remove filesystem for e6054e6bd8740f5a2a5d9bb025ff785bcb5249bc62cdc33a76fad2dfb77f341e: remove /opt/io1/docker/containers/e6054e6bd8740f5a2a5d9bb025ff785bcb5249bc62cdc33a76fad2dfb77f341e/shm: device or resource busy
...
```

**Comment by piontec** [comment]
Most of the time, we got very low ps/run/info response times (< 0.5s), but still sometimes 'docker ps' is not finished within 5s.

**Comment by mlaventure** [comment]
@piontec when they get stuck, could you send a `SIGUSR` to both the `docker-containerd` and `dockerd` process so we can get the stacktraces?

**Comment by piontec** [comment]
@mlaventure This happened again, exactly the same problem, but this time way less docker objects. Please find `docker info` below and thread dump of both dockerd and docker-containerd in attachment
```
# docker info
Containers: 359
 Running: 322
 Paused: 0
 Stopped: 37
Images: 563
Server Version: 1.12.6-cs7
Storage Driver: aufs
 Root Dir: /opt/io1/docker/aufs
 Backing Filesystem: extfs
 Dirs: 3211
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay host bridge null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-59-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 128
Total Memory: 1.876 TiB
Name: ip-10-69-11-89
ID: UGZS:UFD3:GB4C:W5MX:JU2L:K7PH:6ZWS:4GPM:27Q5:UNNN:X3DC:YDT7
Docker Root Dir: /opt/io1/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 4644
 Goroutines: 3030
 System Time: 2017-02-10T08:19:27.173271089Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```
[docker-dump-2017-02-10.txt](https://github.com/docker/docker/files/766093/docker-dump-2017-02-10.txt)

**Comment by tonistiigi** [comment]
@mlaventure 

Seems to be blocked in

```
goroutine 792992 [semacquire, 70 minutes]:
sync.runtime_Semacquire(0xc8204021fc)
\t/usr/local/go/src/runtime/sema.go:47 +0x26
sync.(*Mutex).Lock(0xc8204021f8)
\t/usr/local/go/src/sync/mutex.go:83 +0x1c4
github.com/docker/docker/layer.(*layerStore).GetMountID(0xc8204021c0, 0xc83891fcc0, 0x40, 0x0, 0x0, 0x0, 0x0)
\t/usr/src/docker/.gopath/src/github.com/docker/docker/layer/layer_store.go:499 +0x64
github.com/docker/docker/daemon.(*Daemon).Cleanup(0xc820478000, 0xc827fa05a0)
\t/usr/src/docker/.gopath/src/github.com/docker/docker/daemon/start.go:179 +0x10c
github.com/docker/docker/daemon.(*Daemon).containerStart.func1(0xc835d330e0, 0xc827fa05a0, 0xc820478000)
\t/usr/src/docker/.gopath/src/github.com/docker/docker/daemon/start.go:114 +0x89
github.com/docker/docker/daemon.(*Daemon).containerStart(0xc820478000, 0xc827fa05a0, 0x7f713ece3028, 0xc828e41510)
\t/usr/src/docker/.gopath/src/github.com/docker/docker/daemon/start.go:163 +0xc98
```
but the goroutine holding the layerstore lock is brand new.
```
goroutine 792235 [runnable]:
strings.Contains(0xc83a1c2780, 0x71, 0xc829bd1a40, 0x59, 0xc83a1c2780)
\t/usr/local/go/src/strings/strings.go:132
github.com/docker/docker/daemon/graphdriver/aufs.debugEBusy(0xc829bd1a40, 0x59, 0x0, 0x0, 0x0, 0x0, 0x0)
\t/usr/src/docker/.gopath/src/github.com/docker/docker/daemon/graphdriver/aufs/aufs.go:293 +0x5e7
github.com/docker/docker/daemon/graphdriver/aufs.(*Driver).Remove(0xc8204a6120, 0xc8321a4c80, 0x40, 0x0, 0x0)
\t/usr/src/docker/.gopath/src/github.com/docker/docker/daemon/graphdriver/aufs/aufs.go:325 +0xdf9
github.com/docker/docker/layer.(*layerStore).ReleaseRWLayer(0xc8204021c0, 0x7f7130a1aa80, 0xc82d7983c8, 0x0, 0x0, 0x0, 0x0, 0x0)
\t/usr/src/docker/.gopath/src/github.com/docker/docker/layer/layer_store.go:526 +0x36a
github.com/docker/docker/daemon.(*Daemon).cleanupContainer(0xc820478000, 0xc82500a5a0, 0x1, 0x0, 0x0)
\t/usr/src/docker/.gopath/src/github.com/docker/docker/daemon/delete.go:129 +0x941
github.com/docker/docker/daemon.(*Daemon).ContainerRm(0xc820478000, 0xc833687eb9, 0x40, 0xc8265b4930, 0x0, 0x0)
\t/usr/src/docker/.gopath/src/github.com/docker/docker/daemon/delete.go:42 +0x220
```

**Comment by mlaventure** [comment]
@tonistiigi I don't think the dump shows the time the current thread has been running for, does it?

`lsof` might be taking a long time to complete.

**Comment by mlaventure** [comment]
@piontec could you check how long the lsof process has been running?

**Comment by tonistiigi** [comment]
@mlaventure Yes, checked the source. The time there is the time goroutine has been blocked, not the creation time. So no runnable goroutine ever has a duration.  The goroutine ID is smaller so I think the `debugEBusy` goroutine is older. So it seems that lsof never returned.

**Comment by piontec** [comment]
Guys, we hit this problem again, our docker host is almost unusable due to this bug. Sometimes it happens every few hours now. Please find below info about running `lsof` processes, thread dump for both `dockerd` and `docker-containerd` and `docker info`.

```
# ps aux | grep lsof
root      90013  0.0  0.0  12944  1012 pts/53   S+   08:44   0:00 grep --color=auto lsof
root     111139  0.0  0.0 294592 276012 pts/88  S+   Feb10   0:33 lsof
root     111164  0.0  0.0  17076   184 pts/88   S+   Feb10   0:00 lsof
root     121565 98.5  8.0 162341916 162328504 ? R    05:23 198:03 lsof
root     121567  0.0  0.0  15052   172 ?        S    05:23   0:00 lsof
[docker-dump-2017-02-13.thrd_dump.txt](https://github.com/docker/docker/files/770533/docker-dump-2017-02-13.thrd_dump.txt)

```

```
# docker info
Containers: 341
 Running: 327
 Paused: 0
 Stopped: 14
Images: 803
Server Version: 1.12.6-cs8
Storage Driver: aufs
 Root Dir: /opt/io1/docker/aufs
 Backing Filesystem: extfs
 Dirs: 3604
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-59-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 128
Total Memory: 1.876 TiB
Name: ip-10-69-11-89
ID: UGZS:UFD3:GB4C:W5MX:JU2L:K7PH:6ZWS:4GPM:27Q5:UNNN:X3DC:YDT7
Docker Root Dir: /opt/io1/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 4661
 Goroutines: 3247
 System Time: 2017-02-13T08:50:30.750715758Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

```
# docker version
Client:
 Version:      1.12.6-cs8
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   880bc7e
 Built:        Wed Feb  8 14:59:44 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6-cs8
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   880bc7e
 Built:        Wed Feb  8 14:59:44 2017
 OS/Arch:      linux/amd64
```

**Comment by mlaventure** [comment]
@piontec the culprit seems to be `121565` it's takin '98.5%' of the cpu too.

It may be an issue with `lsof` since I doubt they would be so many open files that it would take more than 3h to process them.

This path of code is only called when the aufs driver cannot unmount a container though.

As a workaround I thing we could add a timeout to the `lsof` execution

**Comment by tonistiigi** [comment]
cc @anusha-ragunathan 

Ideally, we shouldn't keep the lock for this debug collection at all if it can take a lot of time.

**Comment by jcberthon** [comment]
Hi everyone,

We are hit by a similar fate. Usually this happens when we do a `docker-compose up -d` which seems to never return, then after `docker ps` and many other commands never return. Only "solution" is to restart the docker.service unit, which is not very cool because all applications running inside the containers are healthy and restarting the service shuts them down and make them restart for nothing.

We have a load well below 1.0 (ca. 0.75), only about 10% of the memory is used, no swap is used.

We do not know if this happens when we do the docker-compose command or if it was already unresponsive before and this command is also blocked just like the other `docker ps`, etc.

Nothing abnormal reported by `journald` when this happens.

Our docker info:

```
Containers: 11
 Running: 9
 Paused: 0
 Stopped: 2
Images: 274
Server Version: 1.13.1
Storage Driver: devicemapper
 Pool Name: data-dockerthinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 18.52 GB
 Data Space Total: 2.302 TB
 Data Space Available: 2.283 TB
 Metadata Space Used: 8.79 MB
 Metadata Space Total: 16.98 GB
 Metadata Space Available: 16.97 GB
 Thin Pool Minimum Free Space: 230.2 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 1
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.6.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 48
Total Memory: 125.6 GiB
Name: spcbo-infra-host-01.vcs.de
ID: (...)
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by mlaventure** [comment]
@jcberthon you're on devicemapper, so it's probably a devicemapper issue, maybe https://github.com/docker/docker/issues/30003 ?

**Comment by jcberthon** [comment]
@mlaventure thanks for the hint, indeed in `dmesg` I have many errors like `device-mapper: thin: Deletion of thin device 2486 failed.` so it could be a `device-mapper` bug, however does not look related to the one you mention. Anyway, I just need to see if those messages correlate (in time) with some of docker commands hanging up. But that's a start, thank you.

**Comment by mlaventure** [comment]
@jcberthon feel free to open a new issue when you get more information if it doesn't match any other open ones

**Comment by piontec** [comment]
@mlaventure Any progress? It happens so frequently, that we had write our own "check-and-restart" script to auto-fix the problem.

**Comment by jcberthon** [comment]
@mlaventure thanks to your hint I found the related bug. And indeed this is due to the device mapper and the RHEL/CentOS kernel, but here is the correct issue: https://github.com/docker/docker/issues/27381#issuecomment-283676898

As reported by Dan Walsh there is a Kernel bug in RHEL until 7.3 which will be addressed in next point release (7.4). Until then, he provides a work around but this also means that one cannot use live restore for the time being.

This comment is just for the record.
